ember-toucan-core/src/components/form/label.hbs

```diff
@@ -0,0 +1,3 @@
+<label class="type-md-tight text-body-and-labels block" ...attributes>
```

Comment:
headless-form automatically applies `for` to labels, and a matching `id` to the control element, at least when using the provided components. This is the most basic a11y rule we need to follow for forms. So I wonder if/how this component should be concerned about...

The user brings their own control component, so we cannot set an `id` for them. So we could either generate an id (e.g. using `unique-id` helper), automatically set it as `for` on the label, and yield it, so users cann apply it to their controls, like this:

```hbs
<Form::Field as |field|>
  <field.Label>First name</field.Label>
  <field.Control>
    <input id={{field.id}} ...attributes />
  </field.Control>
  <field.Hint>Hint text below the input</field.Hint>
</Form::Field>
```

Or we choose to not make this our concern, but then we should at least call this out in the docs!

For the more opinionated toucan-forms, I think we will wire this all up correctly for the user, right?

Reply:
Definitely! Thank you for bringing this up as a whole.  Yeah, I think we'll need to figure this out.  @nicolechung and I were discussing this last week as we were pairing.  We'll need to do the `id` + `for`, as well as "link" the hint + error blocks as well via something like `aria-describedby` or similar.  We're still discussing if we do this work as part of _this_ PR or a follow up one.  Our goal at the moment is pairing to get this far enough along that we can work on separate tasks without stepping on each other's toes.  We're getting close!

> The user brings their own control component, so we cannot set an id for them. So we could either generate an id (e.g. using unique-id helper), automatically set it as for on the label, and yield it, so users cann apply it to their controls, like this

Yeah!  I like this idea a lot 😄.  Commit coming soon!

> Or we choose to not make this our concern, but then we should at least call this out in the docs!

Absolutely! If we decide it's not our concern here, we can call it out in the docs.  I'd definitely _prefer_ if we can do as much as we can though! (but maybe separate PR for wiring things up properly?)

> For the more opinionated toucan-forms, I think we will wire this all up correctly for the user, right?

Oh yeah, definitely!

Reply:
Okay, yeah, after playing around a bit tonight it's a bit tricky as we don't have access directly to the control the user renders.  Due to that, I think your example above is the best we can do.

Here's what we can provide to make things fully accessible, so that the hint + error blocks get read to screenreaders:

```hbs
{{#let (unique-id) as |uniqueId|}}
  {{#let (concat "hint-" uniqueId) as |hintId|}}
    {{#let (concat "error-" uniqueId) as |errorId|}}
      {{yield
        (hash
          Label=this.Label
          Hint=this.Hint
          Control=this.Control
          Error=this.Error
          id=uniqueId
          hintId=hintId
          errorId=errorId
        )
      }}
    {{/let}}
  {{/let}}
{{/let}}
```

And then usage:

```hbs
<Form::Field as |field|>
  <field.Label for={{field.id}}>First name</field.Label>
  <field.Hint id={{field.hintId}}>What should we call you?</field.Hint>
  <field.Control>
    <input
      class='border bg-basement text-titles-and-attributes'
      id={{field.id}}
      aria-describedby='{{field.hintId}} {{field.errorId}}'
      ...attributes
    />
  </field.Control>
  <field.Error id={{field.errorId}}>error</field.Error>
</Form::Field>
```

Since we can't access the control element directly, as that's provided by the consumer, we _can_ provide IDs for consumers to wire things up as they please.  Then they can put the `for` + `id` + `aria-describedby` in the correct places.  What's your feeling about something like this?

Reply:
Yeah, seems good overall.

Some (minor) notes:

* we could wire up the `for` attribute on the label already, as we do have control over the `<label>` element. headless-form does this. But not sure if that's really preferable here. We have two sides of the same coin here, `for` on the label, and `id` on the control. But auto-applying `for` would obfuscate the one, will leaving the other for the user to do. So maybe letting the user do both is less confusing? I'm 50/50 on this one, so your call! :wink:
* FWIW, I used `aria-errormessage` for linking from the control to the element holding the errors., instead of `aria-describedby` with multiple IDs. Not sure which is better, or if there is any significant difference. But "errormessage" sounded more specific, so I chose that one...
* you don't have to use those nested `{{#let}}`, you can also introduce multiple new local vars in a single one, like [here](https://github.com/CrowdStrike/ember-headless-form/blob/main/packages/ember-headless-form/src/components/-private/field.hbs#L1-L7)

Reply:
Appreciate all of the notes! This is very helpful.   I like the idea of using `aria-errormessage` + `aria-invalid="true"`.  That seems a bit more explicit for error messages and then we can use `aria-describedby` for helper text only!

Reply:
Alright, I gave `aria-errormessage` a shot but ran into something similar to https://github.com/nvaccess/nvda/issues/8318 with voice over as well. The screenreader wouldn't announce the error message, but had no problems with the hint text.  I added two examples of wiring this up in the docs - one using `aria-describedby`  and the other using `aria-errormessage`.  If we can get `aria-errormesssage` to announce properly, I say we roll with that in our Form fields!

Reply:
Hm, that's bad. Thanks for testing this, I should do the same for headless btw...

Probably have to change this to `describedby` as well in headless-form, tracking here: https://github.com/CrowdStrike/ember-headless-form/issues/47

Reminds me of the old days of browser incompatibilities, where you have nice new features in the spec, but you cannot use them due to IE support, instead going for crazy workarounds... 😩